presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -634,10 +636,11 @@ private void appendData(DecodedBlockNode[] decodedBlocks, int fixedWidthRowSize,
             int offset = 0;
             do {
                 int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset);
+                boolean lastBatchInPage = offset + batchSize >= positionCount;
 
                 for (int i = 0; i < channelCount; i++) {
                     blockEncodingBuffers[i].setNextBatch(offset, batchSize);
-                    blockEncodingBuffers[i].appendDataInBatch();
+                    blockEncodingBuffers[i].appendDataInBatch(lastBatchInPage);
```

Comment:
I'm wondering if it would be clearer to add separate API to signal that there will be no more batches and call it at the same time you call allocator.returnArray(serializedRowSizes). Also, it would be helpful to update documentation for BlockEncodingBuffer to explain in which order the methods are being called.

```
int[] serializedRowSizes = ensureCapacityWithAllocator(null, positionCount, SMALL, INITIALIZE, allocator);
            try {

                populateSerializedRowSizes(fixedWidthRowSize, variableWidthChannels, serializedRowSizes);

                // Due to the limitation of buffer size, we append the data batch by batch
                int offset = 0;
                do {
                    int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset, serializedRowSizes);

                    for (int i = 0; i < channelCount; i++) {
                        blockEncodingBuffers[i].setNextBatch(offset, batchSize);
                        blockEncodingBuffers[i].appendDataInBatch();
                    }

                    bufferedRowCount += batchSize;
                    offset += batchSize;

                    if (bufferFull) {
                        flush(outputBuffer);
                        bufferFull = false;
                    }
                }
                while (offset < positionCount);
            }
            finally {
                // Return the borrowed array for serializedRowSizes when the current page will be finished in this batch.
                // At this point the batchSize was already calculated and we don't need the serializedRowSizes any more.
                allocator.returnArray(serializedRowSizes);
                for (int i = 0; i < channelCount; i++) {
                    blockEncodingBuffers[i].noMoreBatches();
                }
            }
```

Reply:
@mbasmanova Masha, thanks for the suggestion. I changed the code as you suggested. If in the future we want to share buffers among blockEncodingBuffers, we will need to do it in the  ` for (int i = 0; i < channelCount; i++)` loop. Related issue https://github.com/prestodb/presto/issues/14274